File: C++/Problems/Problems/problem_3301.cpp
#include "problem_3301.h"

bool problem_3301::test() {
    std::vector maximumHeight { 2, 3, 4, 3 };

    int expected = 10;

    int result = maximumTotalSum(maximumHeight);

    std::cout << result << std::endl;

    return result == expected;
}

long long problem_3301::maximumTotalSum(std::vector<int> &maximumHeight) {
    int n = maximumHeight.size();
    std::sort(maximumHeight.begin(), maximumHeight.end());
    int threshold = INT_MAX;
    long result = 0;
    for (int i = n - 1; i >= 0; --i)
    {
        if (threshold < 1)
            return -1;
        int curr = std::min(maximumHeight[i], threshold);
        result += curr;
        threshold = curr - 1;
    }

    return result;
}

File: Python/Problems/Problems/problem3301.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem3301(Problem):
    def test(self) -> bool:
        maximumHeight = [2, 3, 4, 3]

        expected = 10

        result = self.maximumTotalSum(maximumHeight)

        print(result)

        return result == expected

    def maximumTotalSum(self, maximumHeight: List[int]) -> int:
        n = len(maximumHeight)
        maximumHeight.sort()
        threshold = sys.maxsize
        result = 0
        for i in range(n - 1, -1, -1):
            if threshold < 1:
                return -1
            curr = min(maximumHeight[i], threshold)
            result += curr
            threshold = curr - 1

        return result